README.md
# ModCon
ModCon is a tool to manipulate the nucleotide sequence surrounding of splice donors, in order to either enhance or silence their usage. It is based on predictions of the HEXplorer tool, which estimates the property of a genomic sequences, to recruit splice regulatory proteins.
<br/><br/>
## Installation and Usage
#### 1. Install dependencies
The ModCon tool was developed as an R package (version ≥ R-4.0.3) using integrated Perl scripts (version ≥ Perl 5.18.1) for MaxEntScan score calculations. Therefore, if necessary please install Perl and R first. If wanted, the script performs parallel HZEI calculation parallel with multiple cores. In total three R packages are required to execute the application, namely “parallel”, “utils” and "data.table".
<br/><br/>
In order to execute the ModCon R-script, please download every file of the ModCon repository for installation or use the devtools option to install R packages from github.<br/>
`library(devtools)`<br/>
`install_github("caggtaagtat/ModCon")`
<br/><br/>
#### 2. Define coding sequence
First, please define the coding sequence of interest which holds the splice donor in question. The first nucleotide of the entered sequence must be the first nucleotide of a codon, in order to correctly read the encoded amino acid sequence.<br/>
`cds <- "AGTAACAGTTAGACCAAAGGATAGA..." # coding sequence holding the SD of interest`
<br/><br/>
#### 3. Enter position of first SD sequence nucleotide
Now, state the position of the first nucleotide of the splice donor sequence within the entered coding sequence, again be redefining the exemplary variable SD_seq_start_position.<br/>
`sdSeqStart <- 402`
<br/><br/>
#### 4. Execute the main ModCon function
Select whether to optimize SD context for SD usage enhancement or silencing. Per default, the script will generate the optimal (or suboptimal) sequence surrounding for SD usage enhancement (or silencing).<br/>
`finalSequence <- ModCon(cds, sdSeqStart)` or <br/>
`finalSequence <- ModCon(cds, sdSeqStart, optimizeContext=FALSE)`
<br/><br/>
#### 5. OPTIONAL: Adjust additional parameters
Additionally, if the circumstances may require adjustment of further parameters, please do so.<br/>
`finalSequence <- ModCon(cds, sdSeqStartPosition, upChangeCodonsIn=16, downChangeCodonsIn=16, optimizeContext="optimalContext", sdMaximalHBS=10, acMaximalMaxent=4, optiRate=100, nGenerations=30, parentSize=300, startParentSize=1000, bestRate=40, semiLuckyRate=20, luckyRate=5, mutationRate=1e-04, nCores=-1)`
<br/><br/>

## Parameter settings
| Parameters             | Default           | Definition  |
|:----------------------:|:-----------------:| :-----------------------------------------------------------------------------|
|cds                     | CDS of interest   | Coding sequence which holds the splice donor in question.             |
|sdSeqStartPosition      | 1001              | Position of the first nucleotide of the splice donor sequence within the entered codong sequence in column CDS                       |
|upChangeCodonsIn        | 16                | Number of codons to be substituted upstream of the splice donor sequence.            |
|downChangeCodonsIn      | 16                | Number of codons to be substituted downstream of the splice donor sequence.              |
|optimizeContext         | TRUE              | Define, whether the script should either adjust the SD surrounding by codon selection, in order to enhance (TRUE, default) or silence (FALSE) splice donor usage.  |
|sdMaximalHBS            | 10                | Splice donors within the output sequence, which show a HBS of X or higher, will be attempted to be degraded by codon selection.                              |
|acMaximalMaxent         | 4                 | Splice acceptors within the output sequence, which show a MaxEntScan score of X or higher, will be attempted to be degraded by codon selection.             |
|optiRate                | 100               | Change the extent the splice site HEXplorer weight of the donor site should be adjusted. |
|nGenerations            | 50                | Maximal numbers of generations to generate, before the loop of sequence recombination is terminated, in case the condition to end the loop will never be reached. |
|parentSize              | 300               | Number of sequences in the parental generation, which comes from sequences of either the initial or the filial sequence population.    |
|startParentSize         | 1000              | Number of sequences which are randomly generated during the generation of the initial parent population.                                  |
|bestRate                | 40                | X % of parental sequences with the highest fitness are selected to make up parts of the reproducing population.                      |
|semiLuckyRate           | 20                | X % of parental sequences are selected to make up parts of the reproducing population, based on a probability which is derived from the sequence fitness.  |
|luckyRate               | 5                 | X % of sequences are randomly selected to make up parts of the reproducing population |
|mutationRate            | 1e-04             | With a chance of X % each codon within a filial sequence can mutate to a different codon, encoding the same amino acid.                |


NEWS
Changes in version 0.99.0 (2018-05-15)
+ Submitted to Bioconductor


R/ModCon.R
## ModCon function
ModCon <- function(cds, sdSeqStartPosition, upChangeCodonsIn=16,
                   downChangeCodonsIn=16, optimizeContext=TRUE,
                   sdMaximalHBS=10, acMaximalMaxent=4, optiRate=100,
                   nGenerations=30, parentSize=300, startParentSize=1000,
                   bestRate=40, semiLuckyRate=20, luckyRate=5,
                   mutationRate=1e-04, nCores=-1) {

  ## Error messages for wrong entries cds
  if (!all(strsplit(cds, "")[[1]] %in% c("a", "c", "g", "t", "G", "C", "T", "A")))
    stop(paste0("ERROR during setting of variable 'cds'. The entered sequence",
                " must be a character string of A C G and T."))

  ## sdSeqStartPosition
  if (sdSeqStartPosition < 1 | sdSeqStartPosition > (nchar(cds) - 10)) {
    stop(paste0("ERROR during setting of variable 'sdSeqStartPosition'. SD",
                " sequence not within cds."))
  }

  ## upChangeCodonsIn
  if (!is.numeric(upChangeCodonsIn) | (upChangeCodonsIn < 4)) {
    stop(paste0("ERROR during setting of variable 'upChangeCodonsIn'. Must",
                " be numeric and greater 4."))
  }

  ## downChangeCodonsIn
  if (!is.numeric(downChangeCodonsIn) | (downChangeCodonsIn < 4)) {
    stop(paste0("ERROR during setting of variable 'downChangeCodonsIn'. Must",
                " be an numeric and greater 4."))
  }

  ## optimizeContext
  if (!is.logical(optimizeContext)) {
    stop(paste0("Setting of the variable 'optimizeContext' not correct. Must",
                " be a logical value."))
  }

  ## sdMaximalHBS
  if (sdMaximalHBS < 1.8 | sdMaximalHBS > 23.8) {
    stop(paste0("ERROR during setting of variable 'sdMaximalHBS'. HBS limit",
                " must be within 1.8 to 23.8."))
  }

  ## sdMaximalHBS
  if (!is.numeric(sdMaximalHBS)) {
    stop(paste0("ERROR during setting of variable 'acMaximalMaxent'.",
                " MaxEntScan score limit must be numeric."))
  }

  ## optiRate
  if (!is.numeric(optiRate) | optiRate < 0) {
    stop(paste0("ERROR during setting of variable 'optiRate'. optiRate",
                " must be numeric and not lower than 0 ."))
  }

  ## nGenerations
  if (!is.numeric(nGenerations) | nGenerations < 2) {
    stop(paste0("ERROR during setting of variable 'nGenerations'. nGenerations",
                " must be numeric and greater 1."))
  }

  ## parentSize
  if (!is.numeric(parentSize) | parentSize < 10) {
    stop(paste0("ERROR during setting of variable 'parentSize'. parentSize",
                " must be numeric and greater 9."))
  }

  ## startParentSize
  if (!is.numeric(startParentSize) | startParentSize < 10) {
    stop(paste0("ERROR during setting of variable 'startParentSize'. startParentSize",
                " must be numeric and greater 9."))
  }

  ## bestRate
  if (!is.numeric(bestRate) | bestRate <= 0) {
    stop(paste0("ERROR during setting of variable 'bestRate'. bestRate must",
                " be numeric and greater 0."))
  }

  ## semiLuckyRate
  if (!is.numeric(semiLuckyRate) | semiLuckyRate <= 0) {
    stop(paste0("ERROR during setting of variable 'semiLuckyRate'. semiLuckyRate",
                " must be numeric and greater 0."))
  }

  ## luckyRate
  if (!is.numeric(luckyRate) | luckyRate <= 0) {
    stop(paste0("ERROR during setting of variable 'luckyRate'. luckyRate",
                " must be numeric and greater 0."))
  }

  ## mutationRate
  if (!is.numeric(mutationRate) | mutationRate < 0 | mutationRate > 1) {
    stop(paste0("ERROR during setting of variable 'mutationRate'. mutationRate",
                " must be numeric and range from 0 to 1."))
  }

  ## Define region of interest
  upChangeCodons <- (upChangeCodonsIn * 3) - 1
  downChangeCodons <- (downChangeCodonsIn * 3)

  ## Define codonsInsert object, to allow differences between lookup table Codons
  ## and insertion table codonsInsert
  codonsInsert <- Codons

  ## If SD context will be optimized to enhance usage
  if (optimizeContext) {

    ## Find position of first nucleotid in first codon upstream of SD
    endPos <- sdSeqStartPosition
    while (endPos%%3 != 0) endPos <- endPos - 1

    ## Select subsequence to optimize upstream of the SD (within ORF)
    inputSeq <- substr(cds, endPos - upChangeCodons, endPos)
    insertStartPos <- endPos - upChangeCodons

    ## Report next step
    message("Maximizing HZEI integral of upstream sequence...")

    ## Maximize HZEI of upstream sequence
    res <- changeSequenceHZEI(inputSeq, increaseHZEI=TRUE, nGenerations=nGenerations,
                              parentSize=parentSize, startParentSize=startParentSize,
                              bestRate=bestRate, semiLuckyRate=semiLuckyRate,
                              luckyRate=luckyRate, mutationRate=mutationRate,
                              optiRate=optiRate, sdMaximalHBS=sdMaximalHBS,
                              acMaximalMaxent=acMaximalMaxent, nCores=nCores)

    ## Saving wildtype and optimized sequence
    before <- inputSeq
    after <- res[[3]]
    upstreamPositiveSeq <- after

    ## Get downstream sequence for adjustment
    endPos <- sdSeqStartPosition + 10
    while (endPos%%3 != 0) endPos <- endPos + 1
    inputSeq <- substr(cds, endPos + 1, endPos + downChangeCodons)
    insertEndPos <- endPos + downChangeCodons

    ## Report next step
    message("Minimizing HZEI integral of downstream sequence ...")

    ## Minimize HZEI of downstream sequence
    res <- changeSequenceHZEI(inputSeq, increaseHZEI=FALSE, nGenerations=nGenerations,
                              parentSize=parentSize, startParentSize=startParentSize,
                              bestRate=bestRate, semiLuckyRate=semiLuckyRate,
                              luckyRate=luckyRate, mutationRate=mutationRate,
                              optiRate=optiRate, sdMaximalHBS=sdMaximalHBS,
                              acMaximalMaxent=acMaximalMaxent, nCores=nCores)

    ## Saving wildtype and optimized sequence
    before <- inputSeq
    after <- res[[3]]
    downstreamPositiveSeq <- after

    ## Get SD sequence
    sdSeq <- substr(cds, (insertStartPos + upChangeCodons + 1), (insertEndPos -
                                                                   (downChangeCodons)))

    # Fuse all sequences to final sequence
    completeSequence <- paste0(upstreamPositiveSeq, sdSeq, downstreamPositiveSeq)
    completeSequence <- strsplit(completeSequence, "")[[1]]
    cdsInput <- cds
    cds <- strsplit(cds, "")[[1]]
    cds[insertStartPos:insertEndPos] <- completeSequence
    cds <- paste(cds, collapse = "")

    ## Report important information about adjustment
    message(paste("Position of first nt of SD seq:", sdSeqStartPosition))
    message(paste("Length of substituted sequences upstream and dowstream:",
                  (upChangeCodons + 1), "and", downChangeCodons, "nt"))

    ## If SD context will be optimized to silence usage
  } else {

    ## Find position of first nucleotid in first codon upstream of SD
    endPos <- sdSeqStartPosition + 10
    while (endPos%%3 != 0) endPos <- endPos + 1

    ## Select subsequence to optimize upstream of the SD (within ORF)
    inputSeq <- substr(cds, endPos + 1, endPos + downChangeCodons)
    insertEndPos <- endPos + downChangeCodons

    ## Report next step
    message("Maximize HZEI integral of downstream sequence...")

    ## Maximize HZEI of downstream sequence
    res <- changeSequenceHZEI(inputSeq, increaseHZEI=TRUE, nGenerations=nGenerations,
                              parentSize=parentSize, startParentSize=startParentSize,
                              bestRate=bestRate, semiLuckyRate=semiLuckyRate,
                              luckyRate=luckyRate, mutationRate=mutationRate,
                              optiRate=optiRate, sdMaximalHBS=sdMaximalHBS,
                              acMaximalMaxent=acMaximalMaxent, nCores=nCores)

    ## Saving wildtype and optimized sequence
    before <- inputSeq
    after <- res[[3]]
    downstreamPositiveSeq <- after

    ## Get downstream sequence for adjustment
    endPos <- sdSeqStartPosition
    while (endPos%%3 != 0) endPos <- endPos - 1
    inputSeq <- substr(cds, endPos - upChangeCodons, endPos)
    insertStartPos <- endPos - upChangeCodons

    ## Report next step
    message("Minimizing HZEI integral of upstream sequence...")

    ## Minimize HZEI of downstream sequence
    res <- changeSequenceHZEI(inputSeq, increaseHZEI=FALSE, nGenerations=nGenerations,
                              parentSize=parentSize, startParentSize=startParentSize,
                              bestRate=bestRate, semiLuckyRate=semiLuckyRate,
                              luckyRate=luckyRate, mutationRate=mutationRate,
                              optiRate=optiRate, sdMaximalHBS=sdMaximalHBS,
                              acMaximalMaxent=acMaximalMaxent, nCores=nCores)

    ## Saving wildtype and optimized sequence
    before <- inputSeq
    after <- res[[3]]
    upstreamPositiveSeq <- after

    ## Get SD sequence
    sdSeq <- substr(cds, (insertStartPos + upChangeCodons + 1), (insertEndPos -
                                                                   (downChangeCodons)))

    # Fuse all sequences to final sequence
    completeSequence <- paste0(upstreamPositiveSeq, sdSeq, downstreamPositiveSeq)
    completeSequence <- strsplit(completeSequence, "")[[1]]
    cdsInput <- cds
    cds <- strsplit(cds, "")[[1]]
    cds[insertStartPos:insertEndPos] <- completeSequence
    cds <- paste(cds, collapse = "")

    ## Report important information about adjustment
    message(paste("Position of first nt of SD seq:", sdSeqStartPosition))
    message(paste("Length of substituted sequences upstream and dowstream:",
                  (upChangeCodons + 1), "and", downChangeCodons, "nt"))

  }

  ## Return full cds with the SD and its adjusted sequence context
  return(cds)
}

